addons/ui/fnc_openLobbyManager.sqf

```diff
@@ -0,0 +1,400 @@
+    params ["_ctrlButtonCollapse"];
+
+    private _display = ctrlParent _ctrlButtonCollapse;
+    private _ctrlSlots = _display displayCtrl IDC_LM_SLOTS;
+
+    private _fnc_collapse = {
+        // Collapsing [] path causes tree to disappear
+        if !(_this isEqualTo []) then {
+            _ctrlSlots tvCollapse _this;
+        };
+
+        for "_i" from 0 to ((_ctrlSlots tvCount _this) - 1) do {
+            _this + [_i] call _fnc_collapse;
+        };
+    };
+
+    [] call _fnc_collapse;
+}];
+
+private _ctrlButtonOK = _display displayCtrl IDC_OK;
+_ctrlButtonOK ctrlAddEventHandler ["ButtonClick", {
+    collect3DENHistory {
+        params ["_ctrlButtonOK"];
+
+        private _display = ctrlParent _ctrlButtonOK;
+        private _ctrlSlots = _display displayCtrl IDC_LM_SLOTS;
+
+        private _groups = [];
+        private _selected = [];
+
+        // set3DENSelected takes flattened array
+        {
+            _selected append _x;
+        } forEach get3DENSelected "";
+
+        for "_sideIndex" from 0 to ((_ctrlSlots tvCount []) - 1) do {
+            for "_groupIndex" from 0 to ((_ctrlSlots tvCount [_sideIndex]) - 1) do {
+                private _groupPath = [_sideIndex, _groupIndex];
+                private _group = _ctrlSlots getVariable (_ctrlSlots tvData _groupPath);
+                _groups pushBack _group;
+
+                private _leader = leader _group;
+                private _multipleUnits = count units _group > 1;
+
+                private _units = [];
+
+                for "_unitIndex" from 0 to ((_ctrlSlots tvCount _groupPath) - 1) do {
+                    private _unitPath = [_sideIndex, _groupIndex, _unitIndex];
+                    private _unit = _ctrlSlots getVariable (_ctrlSlots tvData _unitPath);
+                    private _description = _unit getVariable QGVAR(description);
+
+                    set3DENSelected [_unit];
+
+                    if (_multipleUnits) then {
+                        do3DENAction "CutUnit";
+                        do3DENAction "PasteUnitOrig";
+                    };
+
+                    _unit = get3DENSelected "Object" select 0;
+                    _units pushBack _unit;
+
+                    add3DENConnection ["Group", _units, _group];
+                    _group selectLeader _leader;
+
+                    _unit set3DENAttribute ["description", _description];
+                };
+            };
+        };
+
+        // Need frame delaying to avoid duplication bugs
+        [_groups, _selected] spawn {
+            params ["_groups", "_selected"];
+
+            private _total = count _groups;
+            startLoadingScreen [localize LSTRING(AdjustingGroupOrder)];
+
+            {
+                private _group = _x;
+                private _callsign = _group getVariable QGVAR(callsign);
+
+                set3DENSelected [_group];
+
+                private _handle = 0 spawn {do3DENAction "CutUnit"};
+                waitUntil {scriptDone _handle};
```

Comment:
Can you explain what this is doing?

Reply:
Delaying the 3DEN actions by a frame (at least) and then resuming after they occur.

Reply:
I see. spawn takes at least one frame, so the following waitUntil waits for at least one frame.

Could still be rewritten in unscheduled with some stacked execNextFrame decrementing some array. Idk what that would add besides proving a point, ... so I will probably do that.

Reply:
I thought `execNextFrame` was not available in 3DEN, that is why I used scheduled.

Reply:
to delay one frame you could also just make the waitUntil return false once. But I guess the code for that would atleast look more complicated.

Reply:
>I thought execNextFrame was not available in 3DEN, that is why I used scheduled.

Yeah. Could add a RscMapControl with Draw eh as each frame engine though. Anything to avoid waitUntil and spawn. lol

Reply:
If we check for bad code we always search for spawn but rarely for waitUntil. Just hide the spawn and noone will know ;)

Reply:
Oh. Didn't see second spawn above.